compiler/qsc_parse/src/lex/raw.rs

```diff
@@ -162,22 +162,34 @@ enum StringKind {
 }
 
 #[derive(Clone, Copy, Debug, Eq, PartialEq, Sequence)]
-pub(crate) enum CommentKind {
+pub enum CommentKind {
     Normal,
     Doc,
 }
 
 #[derive(Clone)]
-pub(super) struct Lexer<'a> {
+pub struct Lexer<'a> {
     chars: Peekable<CharIndices<'a>>,
     interpolation: u8,
+    starting_offset: u32,
 }
 
 impl<'a> Lexer<'a> {
-    pub(super) fn new(input: &'a str) -> Self {
+    #[must_use]
+    pub fn new(input: &'a str) -> Self {
         Self {
             chars: input.char_indices().peekable(),
             interpolation: 0,
+            starting_offset: 0,
+        }
+    }
+
+    #[must_use]
+    pub fn new_with_starting_offset(input: &'a str, starting_offset: u32) -> Self {
```

Comment:
Hmm, do you really need this additional method for this use case? You could just adjust the starting offset when you pass in the string slice:

```rs
`Lexer::new(&input[offset..])
```

Feels much cleaner to me.

Reply:
The issue is that the raw tokens won't be created with the proper offsets for their spans, because to this instantiation of the Lexer, the beginning of the slice is the beginning of the whole user code. This is why it is necessary to have some way of telling the Lexer to start creating token offsets at a value other than zero.

Reply:
Sorry, yeah I got this backwards. In that case, my suggestion would be to add the offset the spans at the very end, when you're creating the Concrete tokens.

After all, this is a problem that we already have in the compiler. The lexer is created with offset zero, but then each source file needs to be offsetted per their SourceMap offset. The current solution in the compiler is to do it _after_ parsing (using the `Offsetter` visitor). Similarly, I think in your implementation, you can also do it at the very end. 

IMO, if the compiler didn't need this extra method on the offset neither should your implementation.

Reply:
Resolved offline - I finally grok the purpose of this method and think it's fine as is